packages/flutter/lib/src/widgets/animated_cross_fade.dart

```diff
@@ -0,0 +1,207 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/rendering.dart';
+import 'package:meta/meta.dart';
+
+import 'animated_size.dart';
+import 'basic.dart';
+import 'framework.dart';
+import 'transitions.dart';
+
+/// Specifies which of the children to show. See [AnimatedCrossFade].
+///
+/// The child that is shown will fade in, and while the other will fade out.
+enum CrossFadeState {
+  /// Show the first child and hide the second.
+  showFirst,
+  /// Show the second child and hide the first.
+  showSecond
+}
+
+/// A widget that cross-fades between two children and animates itself between
+/// their sizes. The animation is controlled through the [crossFadeState]
+/// parameter. [firstCurve] and [secondCurve] represent the opacity curves of
+/// the two children, and will typically be [Interval] objects. Note that
+/// [firstCurve] is inverted, i.e. it fades out when providing a growing curve
+/// like [Curves.linear]. [sizeCurve] is the curve used to animated between the
+/// size of the fading out child and the size of the fading in child.
+///
+/// In the case where the two children have different sizes, the animation crops
+/// overflowing children during the animation by aligning them to the top.
+/// This means that the bottom will be clipped.
+class AnimatedCrossFade extends StatefulWidget {
+  /// Creates a cross fade animation widget.
+  ///
+  /// The [duration] of the animation is the same for all components (fade in,
+  /// fade out, and size), and you can pass [Interval]s instead of [Curve]s in
+  /// order to have finer control.
+  AnimatedCrossFade({
+    Key key,
+    this.firstChild,
+    this.secondChild,
+    this.firstCurve: Curves.linear,
+    this.secondCurve: Curves.linear,
+    this.sizeCurve: Curves.linear,
+    @required this.crossFadeState,
+    @required this.duration
+  }) : super(key: key) {
+    assert(this.firstCurve != null);
+    assert(this.secondCurve != null);
+    assert(this.sizeCurve != null);
+  }
+
+  /// The first child. It fades in an out according to [crossFadeState].
+  final Widget firstChild;
```

Comment:
It's not really that helpful to explain that firstChild is the "first child".  You might as well explain that for CrossFadeState.showFirst the firstChild will be faded in.